File: src/pcl-bindings/Convolution3DGaussianKernelPointXYZPointXYZ.cpp
#include "export.h"
#include "pcl/pcl_base.h"
#include "pcl/point_types.h"
#include <pcl/filters/convolution_3d.h>
#include <memory>

using namespace pcl;
using namespace std;
using namespace Eigen;
using namespace filters;

using point_t = PointXYZ;
using pointcloud_t = PointCloud<point_t>;
using kernel_t = GaussianKernel<point_t, point_t>;
using convolution_t = Convolution3D<point_t, point_t, kernel_t>;

EXPORT(convolution_t*) convolution_3d_gaussian_kernel_pointxyz_pointxyz_ctor()
{
    return new convolution_t();
}

EXPORT(void) convolution_3d_gaussian_kernel_pointxyz_pointxyz_delete(convolution_t** ptr)
{
    delete* ptr;
    *ptr = NULL;
}

EXPORT(void) convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_input_cloud(convolution_t* ptr, pointcloud_t* cloud)
{
    pointcloud_t::Ptr shared = std::make_shared<pointcloud_t>(*cloud);
    ptr->setInputCloud(shared);
}

EXPORT(void) convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_kernel(convolution_t* ptr, kernel_t* kernel)
{
    ptr->setKernel(*kernel);
}

EXPORT(void) convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_radius_search(convolution_t* ptr, double radius)
{
    ptr->setRadiusSearch(radius);
}

EXPORT(double) convolution_3d_gaussian_kernel_pointxyz_pointxyz_get_radius_search(convolution_t* ptr, double radius)
{
    return ptr->getRadiusSearch();
}

EXPORT(void) convolution_3d_gaussian_kernel_pointxyz_pointxyz_convolve(convolution_t* ptr, pointcloud_t* output)
{
    ptr->convolve(*output);
}

EXPORT(void) convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_indices_vector(convolution_t* ptr, std::vector<int>* indices)
{
    pcl::PointIndices::Ptr indices_ptr(new pcl::PointIndices());
    indices_ptr->indices = *indices;
    ptr->setIndices(indices_ptr);
}

EXPORT(void) convolution_3d_gaussian_kernel_pointxyz_pointxyz_get_indices_vector(convolution_t* ptr, std::vector<int>* indices)
{
    pcl::IndicesPtr indices_ptr = ptr->getIndices();
    indices->assign(indices_ptr->begin(), indices_ptr->end());
}

EXPORT(void) convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_indices(convolution_t* ptr, std::size_t row_start, std::size_t col_start, std::size_t nb_rows, std::size_t nb_cols)
{
    ptr->setIndices(row_start, col_start, nb_rows, nb_cols);
}

File: src/Pcl.NET/Convolution3DGaussianKernelPointXYZPointXYZ.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Pcl.NET
{
    public class Convolution3DGaussianKernelPointXYZPointXYZ : Convolution3D<PointXYZ, PointXYZ, GaussianKernel<PointXYZ, PointXYZ>>
    {
        private readonly VectorInt _indices;
        private PointCloud<PointXYZ>? _input = null;
        private GaussianKernel<PointXYZ, PointXYZ>? _kernel = null;
        private int _numberOfThreads;

        public override int NumberOfThreads
        {
            get
            {
                return _numberOfThreads;
            }
            set
            {
                _numberOfThreads = value;
                Invoke.convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_threads(_ptr, value);
            }
        }
        public override GaussianKernel<PointXYZ, PointXYZ>? Kernel
        {
            get
            {
                ThrowIfDisposed();
                return _kernel;
            }
            set
            {
                ThrowIfDisposed();
                _kernel = value;
                Invoke.convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_kernel(_ptr, _kernel);
            }
        }
        public override PointCloud<PointXYZ>? Input
        {
            get
            {
                ThrowIfDisposed();
                return _input;
            }
            set
            {
                ThrowIfDisposed();
                _input = value;
                Invoke.convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_input_cloud(_ptr, _input);
            }
        }
        public override VectorInt Indices
        {
            get
            {
                ThrowIfDisposed();
                Invoke.convolution_3d_gaussian_kernel_pointxyz_pointxyz_get_indices_vector(_ptr, _indices);
                return _indices;
            }
            set
            {
                ThrowIfDisposed();
                Invoke.convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_indices_vector(_ptr, value);
            }
        }
        /// <summary>
        /// Initializes a new instance of the <see cref="Convolution3DGaussianKernelPointXYZPointXYZ"/> class.
        /// </summary>
        public Convolution3DGaussianKernelPointXYZPointXYZ()
        {
            _indices = new VectorInt();
            _ptr = Invoke.convolution_3d_gaussian_kernel_pointxyz_pointxyz_ctor();
        }
        
        public override PointCloud<PointXYZ> Convolve()
        {
            PointCloudXYZ output = new PointCloudXYZ();
            Invoke.convolution_3d_gaussian_kernel_pointxyz_pointxyz_convolve(_ptr, output);
            return output;
        }

        public override void SetIndices(long row_start, long col_start, long nb_rows, long nb_cols)
        {
            Invoke.convolution_3d_gaussian_kernel_pointxyz_pointxyz_set_indices(_ptr, row_start, col_start, nb_rows, nb_cols);
        }

        protected override void DisposeObject()
        {
            if (!_suppressDispose)
            {
                Invoke.convolution_3d_gaussian_kernel_pointxyz_pointxyz_delete(ref _ptr);
                _ptr = IntPtr.Zero;
            }
        }
    }
}
